pco_python/src/lib.rs

```diff
@@ -0,0 +1,76 @@
+use numpy::{Element, PyArrayDyn};
+use pyo3::exceptions::PyRuntimeError;
+use pyo3::prelude::{pymodule, FromPyObject, PyModule, PyObject, PyResult, Python};
+use pyo3::types::PyBytes;
+
+use pco::data_types::NumberLike;
+use pco::standalone::{auto_compress, simple_decompress_into};
+use pco::DEFAULT_COMPRESSION_LEVEL;
+
+// The Numpy crate recommends using this type of enum to write functions that accept different Numpy dtypes
+// https://github.com/PyO3/rust-numpy/blob/32740b33ec55ef0b7ebec726288665837722841d/examples/simple/src/lib.rs#L113
+#[derive(FromPyObject)]
+enum ArrayDynFloat<'py> {
+  F32(&'py PyArrayDyn<f32>),
+  F64(&'py PyArrayDyn<f64>),
+  I32(&'py PyArrayDyn<i32>),
+  I64(&'py PyArrayDyn<i64>),
+  U32(&'py PyArrayDyn<u32>),
+  U64(&'py PyArrayDyn<u64>),
+}
+
+fn compress_typed<T: NumberLike + Element>(py: Python, arr: &PyArrayDyn<T>) -> PyResult<PyObject> {
+  let arr_ro = arr.readonly();
+  let src = arr_ro.as_slice()?;
+  let compressed = auto_compress(src, DEFAULT_COMPRESSION_LEVEL);
+  Ok(PyBytes::new(py, &compressed).into())
+}
+
+fn decompress_typed<T: NumberLike + Element>(
+  compressed: &PyBytes,
+  out: &PyArrayDyn<T>,
+) -> PyResult<()> {
+  let mut out_rw = out.readwrite();
+  let dst = out_rw.as_slice_mut()?;
+  let src = compressed.as_bytes();
+  let progress = simple_decompress_into(src, dst)
+    .map_err(|e| PyRuntimeError::new_err(format!("pco decompression error: {}", e)))?;
+  if !progress.finished {
+    Err(PyRuntimeError::new_err(
+      "decompression didn't finish. Buffer too small?",
+    ))
+  } else {
+    Ok(())
+  }
+}
+
+#[pymodule]
+fn pcodec(_py: Python<'_>, m: &PyModule) -> PyResult<()> {
+  m.add("__version__", env!("CARGO_PKG_VERSION"))?;
```

Comment:
Wouldn't it be simpler to keep these versions always the same? Why would we need to have separate versions?

:+1: to a README for the Python package anyway.

Reply:
I'm of the opinion that we should have separate versions, each entirely based on API compatibility.

My arguments:
* There's a high chance we'll want to make API-breaking changes to the Rust crate without breaking the Python package and forcing people to bump their semver. I wouldn't want people lagging behind on their Python package's internals because they need to explicitly do a no-op upgrade.
* There's also a chance (albeit lower for now) we'll want to make an API-breaking change to the Python package without any corresponding change to the Rust crate's API. 

Here are examples of packages I know about and how they deal with this problem:
* [The Rust zstd crate](https://github.com/gyscos/zstd-rs) had a versioning scheme of a.b.c+d.e.f where d.e.f is the original C++ implementation's version. They recently changed this to just a.b.c. I'm guessing this just saves them a bit of effort? The +d.e.f doesn't add much if you can look up the corresponding version in documentation when you need it.
* [The Rust numpy crate](https://github.com/PyO3/rust-numpy) has versions unrelated to Python numpy's.
* [The family of Rust Arrow crates](https://github.com/apache/arrow-rs) uses a single version for all of their many packages. [They admit this is a historical oddity](https://github.com/apache/arrow-rs/blob/master/arrow/README.md#versioning--releases), though. I find it really annoying to use these crates because they bump major version all the time. I don't think pco will ever be as big, but I'd like to avoid this malady.